amp.php

```diff
@@ -133,8 +133,28 @@ function amp_init() {
 // Make sure the `amp` query var has an explicit value.
 // Avoids issues when filtering the deprecated `query_string` hook.
 function amp_force_query_var_value( $query_vars ) {
-	if ( isset( $query_vars[ amp_get_slug() ] ) && '' === $query_vars[ amp_get_slug() ] ) {
-		$query_vars[ amp_get_slug() ] = 1;
+
+	if ( isset( $query_vars[ AMP_QUERY_VAR ] ) ) {
+		if ( '' === $query_vars[ AMP_QUERY_VAR ] ) {
+			$query_vars[ AMP_QUERY_VAR ] = 1;
+
+			// Allow for some fuzziness on string/integer type matching.
+		} else if ( '1' !== $query_vars[ AMP_QUERY_VAR ] && 1 !== $query_vars[ AMP_QUERY_VAR ] ) {
+			global $wp_rewrite;
+
+			// If there is something after /amp/, pretend we didn't hit that rewrite rule.
+			if ( $wp_rewrite->use_trailing_slashes && '' !== $query_vars[ AMP_QUERY_VAR ] ) {
+
+				// Check for extra-long slugs, which will truncate and remove the /amp/* part.
+				if ( strlen( $query_vars['name'] ) > 190 ) {
+					// Give WP_Query an improbable string to get a 404.
+					$query_vars['name'] = substr( $query_vars['name'], 0, - 20 ) . substr( md5( $query_vars['name'] ), 0, 20 );
+				} else {
+					$query_vars['name'] = $query_vars['name'] . "/amp/" . $query_vars[ AMP_QUERY_VAR ];
```

Comment:
Note that `"/amp/"` here doesn't account for `AMP_QUERY_VAR` being set to something other than `amp`.